api/tests/src/task/notifications/test_opportunity_notification.py

```diff
@@ -4,178 +4,279 @@
+    ):
+        """Test that latest opportunity version is collected for each saved opportunity"""
+        # create a different user
+        user_2 = factories.UserFactory.create()
+        user_2 = link_user_with_email(user_2)
+
+        # Create a saved opportunity that needs notification
+        opp_1 = factories.OpportunityFactory.create(category=OpportunityCategory.OTHER)
+        opp_2 = factories.OpportunityFactory.create(category=OpportunityCategory.OTHER)
+        opp_3 = factories.OpportunityFactory.create(category=OpportunityCategory.OTHER)
+
+        # create old versions  for opps
+        factories.OpportunityVersionFactory.create(
+            opportunity=opp_1,
+        )
+        factories.OpportunityVersionFactory.create(
+            opportunity=opp_2,
+        )
+        factories.OpportunityVersionFactory.create(
+            opportunity=opp_3,
+        )
+
+        # User saved opportunity records
+        factories.UserSavedOpportunityFactory.create(
+            user=user,
+            opportunity=opp_1,
+        )
+        factories.UserSavedOpportunityFactory.create(
+            user=user,
+            opportunity=opp_2,
+        )
+        factories.UserSavedOpportunityFactory.create(
+            user=user_2,
+            opportunity=opp_1,
+        )
+        factories.UserSavedOpportunityFactory.create(
+            user=user_2,
+            opportunity=opp_3,
+        )
+
+        # create new versions for opps
+        factories.OpportunityVersionFactory.create(
+            opportunity=opp_1, created_at=opp_1.created_at + timedelta(minutes=60)
+        )
+        opp_1_v_2 = factories.OpportunityVersionFactory.create(
+            opportunity=opp_1, created_at=opp_1.created_at + timedelta(minutes=160)
+        )
+        opp_2_v_1 = factories.OpportunityVersionFactory.create(
+            opportunity=opp_2, created_at=opp_2.created_at + timedelta(minutes=60)
+        )
+        factories.OpportunityVersionFactory.create(
+            opportunity=opp_3, created_at=opp_3.created_at + timedelta(minutes=60)
+        )
+        opp_3_v_2 = factories.OpportunityVersionFactory.create(
+            opportunity=opp_3, created_at=opp_3.created_at + timedelta(minutes=80)
+        )
+
+        _clear_mock_responses()
+
+        # Instantiate the task
+        task = OpportunityNotificationTask(db_session=db_session)
+
+        results = task._get_latest_opportunity_versions()
+
+        # assert that only the latest version is picked up for each user_saved_opportunity
+        assert len(results) == 4
+
+        for user_saved_opp, latest_opp_ver in results:
+            opp_id = user_saved_opp.opportunity_id
+
+            if opp_id == opp_1.opportunity_id:
+                assert latest_opp_ver == opp_1_v_2
+            elif opp_id == opp_2.opportunity_id:
+                assert latest_opp_ver == opp_2_v_1
+            elif opp_id == opp_3.opportunity_id:
+                assert latest_opp_ver == opp_3_v_2
```

Comment:
I'd suggest comparing ids rather than doing `lastest_opp_ver == <some DB model>` - not sure that behavior is defined. Just use the primary key ID instead